text/router-enhancements.md

```diff
@@ -0,0 +1,211 @@
+no temperature scale specified, the app will detect the best one based on the location of the
+device.
+
+On that situation, `/forecast` is a valid URL and has to point to the same route as the others with
+specific scale.
+
+```js
+const allowedScales = ['celsius', 'fahrenheit', 'kelvin'];
+
+function isAllowedScale(scaleName) {
+  return allowedScales.includes(scaleName);
+}
+
+Router.map(function() {
+  this.route('forecast', { path: 'forecast(/:scale)', constraints: { scale: isAllowedScale } });
+});
+
+
+router.serialize('/forecast/celsius'); // => routeName='forecast', params: { scale: 'celsius' }
+router.serialize('/forecast/newton'); // => not recognized route.
+router.serialize('/forecast'); // => routeName='forecast', params: { }
+```
+
+On this example `forecast` route is invoked even if the scale is ommited, but if it's not ommited,
+the constraint ensures that it's among a white-list of supported scales.
+
+# How We Teach This
+
+Expand the guides on the routing topic to cover this use cases. For devs with Rails, Django or ASP.net
+background this feature should look very familiar already.
+
+# Drawbacks
+
+It adds some extra complexity to the routes for a feature that we've survived without for years.
+
+Particularly the dynamic segment constraints, without considering the case of lazy-loaded engines,
+can be emulated without too much effort today by manually checking the value of the dynamic segments
+in the routes, so this proposal adds some vector of complexity since it's not clear which approach
+to use as they admittedly overlap.
+
+# Alternatives
+
+What other designs have been considered? What is the impact of not doing this?
+
+# Unresolved questions
+
+- What arguments should the constraint function receive apart from the dynamic segment itself? The
+route seems a reasonable request.
+- In routes with more than one dynamic segment, can constraints perform checks on all segments
+at once? Per example:
+
+```js
+Router.map(function() {
+  this.route('activities', {
+    path: 'activities/from/:from/to/:to',
+    constraints: function(from, to) {
+      // This constraint takes both dynamic segments and performs an assertion
+      // that takes both segments into consideration.
+      return moment(from).isBefore(moment(to));
+    }
+  });
+});
+
+
+router.serialize('/activities/from/2016-04-11/to/2016-05-20'); // => matches to `activities`
+router.serialize('/activities/from/2016-06-11/to/2016-05-20'); // => fails to match
+```
+
+- How complex can constraints be? Just pure functions or can they have access to services? I think
+that for simplicity they should start as simple as possible, and they could grow to support more
+complex usages.
+- When an optional dynamic segments are combined with constraints on that segment, and the segment
+is not present, does the constraint function run with `undefined` as value or it doesn't run at all?
+I lean towards the second.
```

Comment:
Kinda weird:

Given these two paths:
```javascript
"/foo/:bar"
"/foo/thing"
"/foo/thing/else"
```

These are the results:
```javascript
deserialize('/foo'); // no constraint check.
deserialize('/foo/thing'); // constraint check.
deserialize('/foo/thing/else'); // constraint check is an implementation detail and may be optimized out.
```

In other words one should never rely upon side effects from a constraint check function (which we're not implementing at this time) as they're only guaranteed to be called if you end up in the route with the constraint check or have multiple dynamic segments at a particular segment count.